src/compiler/ruby_generator.cc

```diff
@@ -37,16 +38,60 @@ using std::vector;
 namespace grpc_ruby_generator {
 namespace {
 
+// returns the full canonical message name
+grpc::string GetCanonicalMessageType(const Descriptor* msg) {
+  const Descriptor* top_level_msg;
+  const FileDescriptor* file_containing_msg;
+  grpc::string resolved_namespace;
+
+  // If msg is nested, find the topmost message
+  top_level_msg = msg;
+  while (top_level_msg->containing_type()) {
+    top_level_msg = top_level_msg->containing_type();
+  }
+  file_containing_msg = top_level_msg->file();
+
+  grpc::string msg_full_name;
+  grpc::string msg_name;
+  grpc::string msg_proto_pkg;
+
+  msg_full_name = top_level_msg->full_name();
+  msg_name = top_level_msg->name();
+  msg_proto_pkg =
+      msg_full_name.substr(0, msg_full_name.length() - msg_name.length());
+
+  if (file_containing_msg->options().has_ruby_package()) {
+    resolved_namespace = file_containing_msg->options().ruby_package();
+  } else {
+    return msg->full_name();
+  }
+
+  // remove trailing period
+  if (msg_proto_pkg.length() > 0) {
+    msg_proto_pkg.pop_back();
+  } else if (resolved_namespace.length() > 0) {
```

Comment:
out of curiosity, do we know if it's actually possible to have a zero-length ruby_package? (I'm fine this is just defensive)

Reply:
Ruby_package could be zero-length if user decides to define ruby_package to be the empty string. It's a weird edge case, but I think we should cover it regardless.

Reply:
This sounds good to me, however, can we add a test case to illustrate?